include/libtorrent/aux_/tracker_list.hpp

```diff
@@ -0,0 +1,115 @@
+/*
+
+Copyright (c) 2022, Arvid Norberg
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+    * Redistributions of source code must retain the above copyright
+      notice, this list of conditions and the following disclaimer.
+    * Redistributions in binary form must reproduce the above copyright
+      notice, this list of conditions and the following disclaimer in
+      the documentation and/or other materials provided with the distribution.
+    * Neither the name of the author nor the names of its
+      contributors may be used to endorse or promote products derived
+      from this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef TORRENT_TRACKER_LIST_HPP_INCLUDED
+#define TORRENT_TRACKER_LIST_HPP_INCLUDED
+
+#include <string>
+#include <unordered_map>
+
+#include "libtorrent/aux_/announce_entry.hpp"
+#include "libtorrent/announce_entry.hpp"
+#include "libtorrent/aux_/vector.hpp"
+#include "libtorrent/aux_/invariant_check.hpp"
+
+
+namespace libtorrent::aux {
+
+struct TORRENT_EXTRA_EXPORT tracker_list
+{
+	using trackers_t = aux::vector<aux::announce_entry>;
+	using iterator = typename trackers_t::iterator;
+	using const_iterator = typename trackers_t::const_iterator;
+
+	aux::announce_entry* find_tracker(std::string const& url);
+
+	// returns true if the tracker was added, and false if it was already
+	// in the tracker list (in which case the source was added to the
+	// entry in the list)
+	bool add_tracker(announce_entry const& ae);
```

Comment:
I haven't had time to review it completely yet. But I remembered one thing that I noticed several times, but all the time I forgot to report. I wonder what is the use of having `announce_entry::source` field? In any case, it seems to be handled incorrectly. Due to the fact that it is not stored in the "fastresume data", it is reset to "magnet" for all trackers after the session is restarted.

Reply:
the only reason is to report it to the client. I would have expected it to say "resume_data" when you restart.

It's similar to reporting whether a torrent has received any incoming connections. It's ephemeral state to indicate whether things appear to be working right. Although, I believe the order in which peers are attempted to be connected to may be affected by which sources we've received them from.

Reply:
> I would have expected it to say "resume_data" when you restart.

This will not make it more useful, since all trackers will have such a source after the first session restart.

Reply:
I was mixing it up with `source` on peers btw. Only trackers that were loaded from resume data will have the source be "magnet", not trackers added manually, nor trackers received via tracker-exchange (but that's not a thing anymore).

Just like whether you've received incoming connections or not isn't saved, this state isn't meant to be saved either. Similar to the source field for peers.

Granted, since tracker-exchange isn't a thing anymore, it's not terribly useful. And since magnet links are deconstructed into the resume data class (`add_torrent_params`), there's no distinction between trackers originating from a magnet link and the resume data.

You can tell whether a tracker was added manually, was part of the .torrent file or came from resume data/magnet link.

Reply:
> You can tell whether a tracker was added manually, was part of the .torrent file or came from resume data/magnet link.

Until the session restarts. Then all the trackers will be "come from resume data". So I think it is useless. Otherwise you should store it in resume data.

Reply:
just as useless as "received incoming connection" is. or the `source` field for peers. The point isn't to remember where it came from, the point is to register that features are still operational.